applications/nrf5340_audio/src/modules/sd_card_playback.c

```diff
@@ -0,0 +1,555 @@
+		if (ret < 0) {
+			LOG_ERR("SD card read err: %d", ret);
+			sd_card_playback_active = false;
+			break;
+		}
+
+		if (lc3_fr_len != lc3_playback_cfg.lc3_frame_length_bytes) {
+			LOG_ERR("Read size (%d) not equal requested size (%d)", lc3_fr_len,
+				lc3_playback_cfg.lc3_frame_length_bytes);
+			ret = -EPERM;
+			break;
+		}
+
+		/* Decode audio data frame */
+		ret = sw_codec_lc3_dec_run(lc3_frame, lc3_playback_cfg.lc3_frame_length_bytes,
+					   pcm_frame_size, decoder_num_ch - 1, pcm_mono_frame,
+					   &pcm_mono_write_size, false);
+		if (ret) {
+			LOG_ERR("Decoding err: %d", ret);
+			sd_card_playback_active = false;
+			break;
+		}
+
+		ret = sd_card_playback_ringbuf_write((char *)pcm_mono_frame, pcm_mono_write_size);
+		if (ret < 0) {
+			LOG_ERR("Load ringbuf err: %d", ret);
+			sd_card_playback_active = false;
+			break;
+		}
+	}
+
+	ret_sd_card_close = sd_card_close(&f_seg_read_entry);
+	/* Check if something inside the for loop failed */
+	if (ret) {
+		LOG_ERR("LC3 playback err: %d", ret);
+		sd_card_playback_active = false;
+		return ret;
+	}
+
+	if (ret_sd_card_close) {
+		LOG_ERR("SD card close err: %d", ret);
+		sd_card_playback_active = false;
+		return ret;
+	}
+
+	sd_card_playback_active = false;
+
+	return 0;
+}
+
+static void sd_card_playback_thread(void *arg1, void *arg2, void *arg3)
+{
+	int ret;
+
+	while (1) {
+		k_sem_take(&m_sem_playback, K_FOREVER);
+		switch (playback_file_format) {
+		case SD_CARD_PLAYBACK_WAV:
+			ret = sd_card_playback_play_wav();
+			if (ret) {
+				LOG_ERR("Wav playback err: %d", ret);
+			}
+
+			break;
+		case SD_CARD_PLAYBACK_LC3:
+			ret = sd_card_playback_play_lc3();
+			if (ret) {
+				LOG_ERR("LC3 playback err: %d", ret);
+			}
+
+			break;
+		}
+	}
+}
+
+bool sd_card_playback_is_active(void)
+{
+	return sd_card_playback_active;
+}
+
+int sd_card_playback_wav(char *filename)
+{
+	if (!sw_codec_is_initialized()) {
+		LOG_ERR("Sw codec not initialized");
+		return -EACCES;
+	}
+
+	playback_file_format = SD_CARD_PLAYBACK_WAV;
+	playback_file_name = filename;
+	k_sem_give(&m_sem_lc3_ringbuf_space_available);
+	k_sem_give(&m_sem_playback);
+
+	return 0;
+}
+
+int sd_card_playback_lc3(char *filename)
+{
+	if (!sw_codec_is_initialized()) {
+		LOG_ERR("Sw codec not initialized");
+		return -EACCES;
+	}
+
+	playback_file_format = SD_CARD_PLAYBACK_LC3;
+	playback_file_name = filename;
+	k_sem_give(&m_sem_lc3_ringbuf_space_available);
+	k_sem_give(&m_sem_playback);
+
+	return 0;
+}
+
+int sd_card_playback_mix_with_stream(void *const pcm_a, size_t pcm_a_size)
+{
+	int ret;
+	uint8_t pcm_b[pcm_frame_size];
+	size_t read_size = pcm_frame_size;
+
+	if (!sd_card_playback_active) {
+		LOG_ERR("SD card playback is not active");
+		return -EACCES;
+	}
+
+	ret = sd_card_playback_ringbuf_read(pcm_b, &read_size);
+	if (ret) {
+		LOG_ERR("Loading data into buffer err: %d", ret);
+		return ret;
+	}
+
+	if (read_size > 0) {
+		ret = pcm_mix(pcm_a, pcm_a_size, pcm_b, pcm_frame_size, B_MONO_INTO_A_STEREO_L);
```

Comment:
This is a segmentation fault waiting to happen. If sd_card_playback_ringbuf_read returns less than pcm_fram_size you will start to mix in data that has not been initialized. I suggest doing:
```suggestion
		ret = pcm_mix(pcm_a, pcm_a_size, pcm_b, read_size, B_MONO_INTO_A_STEREO_L);
```

If sd_card_playback_ringbuf_read returns less than pcm_frame_size you would get an error, but you don't check the return value in audio_datapath.c